openmc/deplete/nuclide.py

```diff
@@ -211,14 +213,190 @@ def to_xml_element(self):
+        """Construct a distribution from a dictionary of yields
+
+        Parameters
+        -----------
+        fission_yields : dict
+            Dictionary ``{energy: {product: yield}}``
+
+        Returns
+        -------
+        FissionYieldDistribution
+        """
+        # mapping {energy: {product: value}}
+        energies = tuple(sorted(fission_yields))
+
+        # Get a consistent set of products to produce a matrix of yields
+        shared_prod = set()
+        for prod_set in map(set, fission_yields.values()):
+            shared_prod |= prod_set
+        ordered_prod = tuple(sorted(shared_prod))
+
+        yield_matrix = empty((len(energies), len(shared_prod)))
+
+        for g_index, energy in enumerate(sorted(energies)):
+            prod_map = fission_yields[energy]
+            for prod_ix, product in enumerate(ordered_prod):
+                yield_val = prod_map.get(product)
+                if yield_val is None:
+                    yield_matrix[g_index, prod_ix] = 0.0
+                else:
+                    yield_matrix[g_index, prod_ix] = yield_val
+
+        return cls(energies, ordered_prod, yield_matrix)
+
+    def to_xml_element(self, root):
+        """Write fission yield data to an xml element
+
+        Parameters
+        ----------
+        root : xml.etree.ElementTree.Element
+            Element to write distribution data to
+        """
+        for energy, yield_obj in self.items():
+            yield_element = ET.SubElement(root, "fission_yields")
+            yield_element.set("energy", str(energy))
+            product_elem = ET.SubElement(yield_element, "products")
+            product_elem.text = " ".join(map(str, yield_obj.products))
+            data_elem = ET.SubElement(yield_element, "data")
+            data_elem.text = " ".join(map(str, yield_obj.yields))
+
+
+class _FissionYield(Mapping):
+    """Mapping for fission yields of a parent at a specific energy
+
+    Separated to support nested dictionary-like behavior for
+    :class:`FissionYieldDistribution`, and allowing math operations
+    on a single vector of yields
+
+    Parameters
+    ----------
+    products : tuple of str
+        Products for this specific distribution
+    yields : numpy.ndarray
+        View into associated :attr:`FissionYieldDistribution.yield_matrix`
+    """
+
+    def __init__(self, products, yields):
+        self.products = products
+        self.yields = yields
+
+    def __getitem__(self, product):
+        if product not in self.products:
+            raise KeyError(product)
+        return self.yields[self.products.index(product)]
+
+    def __len__(self):
+        return len(self.products)
+
+    def __iter__(self):
+        return iter(self.products)
+
+    def __iadd__(self, other):
+        """Increment value from other fission yield"""
+        self.yields += other.yields
+        return self
+
+    def __mul__(self, value):
+        return _FissionYield(self.products, self.yields * value)
```

Comment:
I would implement `__add__` that uses `__iadd__`, and also implement `__imul__` and then change `__mul__` to use it. For completeness, you'll probably want `__rmul__` as well.

Reply:
`__rmul__` will allow you to write the more natural `fraction * yield`